cmd/ci-secret-bootstrap/main_test.go

```diff
@@ -2658,3 +2659,635 @@ func (f *fakeVaultClient) ListKVRecursively(prefix string) ([]string, error) {
+				"cluster-1": {
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "prod-secret-1", Namespace: "namespace-1", Labels: map[string]string{"dptp.openshift.io/requester": "ci-secret-bootstrap"}},
+						Data: map[string][]byte{
+							"key-name-1":    []byte("secret-data"),
+							"merge-with-me": []byte("old-secret-data"),
+						},
+						Type: coreapi.SecretTypeOpaque,
+					},
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "some-name", Namespace: "namespace-1", Labels: map[string]string{"dptp.openshift.io/requester": "ci-secret-bootstrap"}},
+						Data: map[string][]byte{
+							"some-data-key":                []byte("a-secret"),
+							"secretsync-vault-source-path": []byte("secret/user-item-1"),
+						},
+						Type: coreapi.SecretTypeOpaque,
+					},
+				},
+				"cluster-2": {
+					{
+						ObjectMeta: metav1.ObjectMeta{Name: "some-name", Namespace: "namespace-1", Labels: map[string]string{"dptp.openshift.io/requester": "ci-secret-bootstrap"}},
+						Data: map[string][]byte{
+							"some-data-key":                []byte("a-secret"),
+							"secretsync-vault-source-path": []byte("secret/user-item-1"),
+							"merge-with-me":                []byte("old-secret-data"),
+						},
+						Type: coreapi.SecretTypeOpaque,
+					},
+				},
+			},
+		},
+		{
+			id: "fails to find secret in vault, error is expected",
+			config: secretbootstrap.Config{
+				Secrets: []secretbootstrap.SecretConfig{
+					{
+						From: map[string]secretbootstrap.BitWardenContext{
+							"key-name-1": {
+								BWItem: "item-name-1",
+								Field:  "field-name-1",
+							},
+						},
+						To: []secretbootstrap.SecretContext{
+							{
+								Cluster:   "cluster-1",
+								Namespace: "namespace-1",
+								Name:      "prod-secret-1",
+							},
+						},
+					},
+				},
+			},
+			secretGetters: map[string]Getter{
+				"cluster-1": fake.NewSimpleClientset().CoreV1(),
+			},
+			vaultObjects: map[string][]vaultObject{
+				"no-one-wants-me": {
+					{
+						field: "field-name-1",
+						data:  []byte("secret-data"),
+					},
+				},
+			},
+			expectedErrors: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.id, func(t *testing.T) {
+			vaultAddr := testhelper.Vault(t)
+
+			o := options{
+				maxConcurrency: 1,
+				force:          tc.force,
+				config:         tc.config,
+				secretsGetters: tc.secretGetters,
+				secrets: secrets.CLIOptions{
+					VaultPrefix: "secret",
+					VaultAddr:   "http://" + vaultAddr,
+					VaultToken:  testhelper.VaultTestingRootToken,
+				},
+			}
+
+			censor := secrets.NewDynamicCensor()
+			c, err := o.secrets.NewClient(&censor)
+			if err != nil {
+				t.Fatal("Failed to create client.")
+			}
+
+			for item, objects := range tc.vaultObjects {
+				for _, object := range objects {
+					if err := c.SetFieldOnItem(item, object.field, object.data); err != nil {
+						t.Fatalf("couldn't populate vault item %s: %v", item, err)
+					}
+				}
+			}
+
+			if tc.initialData != nil {
```

Comment:
Sorry, but this doesn't make any sense. You already pass fakeclient arounds that can be prepopulated, there is absolutely no reason to have another `initialData` field.

Regardless of that as I explained in my earlier comments, please do _not_ add update tests in your existing `testCases` slice. Instead, iterate twice over it, once for the `Create` case and once for the `Update` case. For the `Update` case you can look at `tc.expectedSecrets` to know which secrets to pre-create. Do so, but manipluate their values to something arbitrary beforehand. I am happy to explain this in more detail over a call, but please do not duplicate all tests with a `Create` and an `Update` variation.

Reply:
There is also the `merge` case, but I've got your point. I'll make the changes. Thanks!

Reply:
I was wrong about the `merge` case, so I removed the initial data and add them when the fake client is being created. I kept the update cases because it doesn't make sense to iterate again to the test cases to test the update case since we control that with the `--force` argument.